### PR TITLE
CB-6253 eliminate custom comparator and introduce prefix matcher

### PR DESCRIPTION
**Pull request by akanto**

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java [hunk 1]
```diff
@@ -85,23 +83,27 @@ private boolean isOsVersionsMatch(Image currentImage, Image newImage) {
...
 
     private Predicate<Image> validateCmVersion(Image currentImage) {
-        return image -> compareVersionStrict(currentImage.getPackageVersions().get(CM_PACKAGE_KEY), image.getPackageVersions().get(CM_PACKAGE_KEY));
+        return image -> upgradePermissionProvider.permitCmAndSatckUpgrade(currentImage.getPackageVersions().get(CM_PACKAGE_KEY),
+                image.getPackageVersions().get(CM_PACKAGE_KEY));
     }
 
     private Predicate<Image> validateStackVersion(Image currentImage) {
-        return image -> compareVersionStrict(currentImage.getPackageVersions().get(STACK_PACKAGE_KEY), image.getPackageVersions().get(STACK_PACKAGE_KEY));
+        return image -> upgradePermissionProvider.permitCmAndSatckUpgrade(currentImage.getPackageVersions().get(STACK_PACKAGE_KEY),
```
Comment (tiborpopovics) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java]: Can you please extract the upgradePermissionProvider.permitCmAndSatckUpgrade call into a new method?

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradePermissionProvider.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import static com.sequenceiq.cloudbreak.cloud.PrefixMatchLength.MINOR;
+
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.cloudbreak.cloud.VersionComparator;
+import com.sequenceiq.cloudbreak.cloud.VersionPrefix;
+
+@Service
+public class UpgradePermissionProvider {
```
[the remark this answers is not in the document]
Comment (akanto) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradePermissionProvider.java]: I used a @Spy in ClusterUpgradeImageFilterTest, but I can definitely add tests for this.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java (hunk 1 above)
Comment (akanto) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java]: Thanks. Done.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradePermissionProvider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (akanto) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradePermissionProvider.java]: Added 100% coverage to it.

**Pull request merged**